### ISSUE TITLE
Crash when running JITaaS with SSL

### ISSUE THREAD

**Issue by dmitry-ten** [eclipse-openj9/openj9#4382]
The crash happens on the server, in `SSLInputStream::Read`, inside `BIO_read(_ssl, buffer, size)`. The latter is a SSL library method.
Stacktrace below:
```
#JITaaS: Server received request for stream 00007F1DCBF00558reading from ssl stream: ssl=0x7f1eac03fd20, buffer=0x7f1fb4006fe0, size=8192
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000004 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
Handler1=00007F1FDB6ECC90 Handler2=00007F1FDAFD3F50 InaccessibleAddress=0000000000000030
RDI=0000000000000000 RSI=00007F1FB4006FE0 RAX=0000000000000000 RBX=00007F1EAC03FD20
RCX=00007F1FB4006FE0 RDX=0000000000002000 R8=00007F1FD82A1700 R9=00007F1FDCDB30FD
R10=000000000000006B R11=0000000000000000 R12=0000000000002000 R13=0000000000000000
R14=00007F1EAC03C240 R15=00007F1FD82B35D0
RIP=00007F1FCD07E974 GS=0000 FS=0000 RSP=00007F1FD829E270
EFlags=0000000000010246 CS=0033 RBP=00007F1FB4006FE0 ERR=0000000000000004
TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000030
xmm0 000000000000ff00 (f: 65280.000000, d: 3.225261e-319)
xmm1 2525252525252525 (f: 623191360.000000, d: 9.532824e-130)
xmm2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm3 ffffffff00000000 (f: 0.000000, d: -nan)
xmm4 2c90d8bdbfcca971 (f: 3217860864.000000, d: 5.047715e-94)
xmm5 08090a0b0c0d0e0f (f: 202182160.000000, d: 5.924543e-270)
xmm6 5bb8398042fadf58 (f: 1123737472.000000, d: 6.877886e+133)
xmm7 aebf7eb216a53c31 (f: 379927616.000000, d: -1.621218e-83)
xmm8 0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm9 3c00000000000000 (f: 0.000000, d: 1.084202e-19)
xmm10 70e84a25bc67f977 (f: 3160930560.000000, d: 7.723029e+235)
xmm11 e9a2c971dcecd2c5 (f: 3706508032.000000, d: -7.190219e+200)
xmm12 0000000000000001 (f: 1.000000, d: 4.940656e-324)
xmm13 08090a0b0c0d0e0f (f: 202182160.000000, d: 5.924543e-270)
xmm14 7fa7d0f972053771 (f: 1912944512.000000, d: 8.362189e+306)
xmm15 7333f363a7b520ca (f: 2813665536.000000, d: 8.718373e+246)
Module=/home/dmitry-ten/xa6480sr6-1206/jre/lib/amd64/compressedrefs/libj9jit29.so
Module_base_address=00007F1FCB919000
Target=2_90_20181204_404088 (Linux 3.10.0-862.11.6.el7.x86_64)
CPU=amd64 (56 logical CPUs) (0xfa8290000 RAM)
----------- Stack Backtrace -----------
(0x00007F1FCD07E974 [libj9jit29.so+0x1765974])
(0x00007F1FCD070232 [libj9jit29.so+0x1757232])
(0x00007F1FCD0B1DF3 [libj9jit29.so+0x1798df3])
_ZN14SSLInputStream4ReadEPvi+0x8d (0x00007F1FCC71013B [libj9jit29.so+0xdf713b])
_ZN6google8protobuf2io25CopyingInputStreamAdaptor4NextEPPKvPi+0xdc (0x00007F1FCCF42776 [libj9jit29.so+0x1629776])
(0x00007F1FCCF3E809 [libj9jit29.so+0x1625809])
_ZN6google8protobuf2io16CodedInputStream7RefreshEv+0x116 (0x00007F1FCCF4076C [libj9jit29.so+0x162776c])
_ZN6google8protobuf2io16CodedInputStreamC2EPNS1_19ZeroCopyInputStreamE+0xdf (0x00007F1FCC4CB90B [libj9jit29.so+0xbb290b])
_ZN6JITaaS8J9Stream12readBlockingINS_15J9ClientMessageEEEvRT_+0x52 (0x00007F1FCC4CCFAC [libj9jit29.so+0xbb3fac])
_ZN6JITaaS14J9ServerStream4readIJmjP8J9MethodP7J9Class10TR_HotnessNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEN2J928IlGeneratorMethodDetailsTypeESt6vectorIP19TR_OpaqueClassBlockSaISH_EESt5tupleIJSC_S3_SH_iSH_SJ_SF_IhSaIhEEbmbjSH_PvSH_SH_SH_mEESC_SC_jEEESK_IJDpT_EEv+0x35 (0x00007F1FCC736D03 [libj9jit29.so+0xe1dd03])
_ZN2TR30CompilationInfoPerThreadRemote12processEntryER21TR_MethodToBeCompiledRN2J917J9SegmentProviderE+0x172 (0x00007F1FCC72945A [libj9jit29.so+0xe1045a])
_ZN2TR24CompilationInfoPerThread14processEntriesEv+0x247 (0x00007F1FCC4DA06D [libj9jit29.so+0xbc106d])
_ZN2TR24CompilationInfoPerThread3runEv+0x60 (0x00007F1FCC4D990E [libj9jit29.so+0xbc090e])
_Z30protectedCompilationThreadProcP13J9PortLibraryPN2TR24CompilationInfoPerThreadE+0x1e6 (0x00007F1FCC4D96CA [libj9jit29.so+0xbc06ca])
(0x00007F1FDAFD4E18 [libj9prt29.so+0x21e18])
_Z21compilationThreadProcPv+0x48d (0x00007F1FCC4D9488 [libj9jit29.so+0xbc0488])
(0x00007F1FDB4422C6 [libj9thr29.so+0xe2c6])
(0x00007F1FDD555DD5 [libpthread.so.0+0x7dd5])
clone+0x6d (0x00007F1FDCE64B3D [libc.so.6+0xfeb3d])
---------------------------------------
```
The crash is intermittent.
At the moment I don't know the reason for this crash. I suspected, that it could be related to not properly freeing `SSL` struct at the end of each compilation, but when I tried freeing it, I got some new crashes, so maybe that's not the issue after all.

**Comment by dmitry-ten** [comment]
@mpirvu fyi

**Comment by AlexeyKhrabrov** [comment]
Here is some info about this issue that might be helpful.

The crash is easily reproducible with DayTrader7 with multiple clients. The immediate cause of the segfault is a NULL pointer to the `SSL` object inside the `BIO` object (discovered using `libopenssl` with debug symbols). Creating/destroying the `SSL` and `BIO` objects in our code appears to be correct. I'm pretty sure this issue is caused by something else.

In all crashing runs I've seen there is memory corruption happening in the malloc heap. This often manifests as aborts in other threads happening around the same time because of internal assertions in `malloc()` or `free()` (e.g. `malloc_consolidate(): invalid chunk size`). Enabling malloc integrity checks (e.g. with the `MALLOC_CHECK_` environment variable) reveals memory corruption detected by `malloc()` calls in other threads (e.g. `malloc(): memory corruption (fast)`).

This memory corruption might not even be directly related to SSL, but is rather just more likely to be triggered when using SSL, possibly because SSL is the only thing (as far as I know) on JITServer that uses relatively long-lived and/or relatively large allocations with `malloc()`. The rest of JITServer code only uses malloc for small objects like some `std::string`s in client session data and short-lived objects during message serialization/deserialization.

**Comment by mpirvu** [comment]
Attn: @samasri

**Comment by mpirvu** [comment]
I cannot reproduce the crashes at the moment, but I do see error messages coming from openssl. E.g.
`139791321061120:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl/record/ssl3_record.c:681:`
The pattern I see is that the client closes the connection (maybe because the compilation queue is empty) and the server experiences an exception when calling `readCompileRequest()` which calls `readMessage()` which calls `readOnceBlocking()` which calls `BIO_read()`. Following the exception we call the destructor of the stream which closes the socket and calls `BIO_free_all(_ssl)`
When the client wants to open a new connection, that connection attempt may fail with the message shown above.
I see that socket descriptors are reused and memory for SSL objects can get reused as well (i.e. `(*OSSL_new)(sslCtx);` may return the same pointer after a short while). This is not bad per se, but I still think that we don't properly destroy everything related to a connection that has ended.

**Comment by mpirvu** [comment]
Interestingly, looking at the openssl code I see that `SSL_set_fd` creates a BIO as well:
```
int SSL_set_fd(SSL *s, int fd)
{
    int ret = 0;
    BIO *bio = NULL;

    bio = BIO_new(BIO_s_socket());

    if (bio == NULL) {
        SSLerr(SSL_F_SSL_SET_FD, ERR_R_BUF_LIB);
        goto err;
    }
    BIO_set_fd(bio, fd, BIO_NOCLOSE);
    SSL_set_bio(s, bio, bio);
    ret = 1;
 err:
    return ret;
}
```
I wonder if we should create a BIO of our own. Our code is around these lines:
```
 SSL *ssl = (*OSSL_new)(sslCtx);
(*OSSL_set_accept_state)(ssl);
(*OSSL_set_fd)(ssl, connfd)
(*OSSL_accept)(ssl)
bio = (*OBIO_new_ssl)(sslCtx, false)
(*OBIO_ctrl)(bio, BIO_C_SET_SSL, true, (char *)ssl)
```

**Comment by dmitry-ten** [comment]
I was able to reproduce the crash with `TestJlmRemoteClassAuth_0` from the `system` test suite.
For me, the crash doesn't happen after a connection failure. It just happens some time into the test, right after a new server stream is initialized, when it first reads a compilation request.

Interestingly, the crash goes away, when I comment out the code for freeing the `BIO` object.

```
 71    virtual ~CommunicationStream()
 72       {
 73       if (_connfd != -1)
 74          close(_connfd);
 75
 76       if (_ssl)
 77          (*OBIO_free_all)(_ssl);
 78       }
```

This indicates to me that either calling `BIO_free_all` is freeing more than it should or our initialization of `BIO` and `SSL` is not correct.

**Comment by mpirvu** [comment]
I too believe this is due to the way we handle BIO creation/destruction. I started to read the SSL implementation to better understand it, but didn't have time to make much progress before going on vacation.

**Comment by dmitry-ten** [comment]
I think I found the invalid memory operation:
When we create an SSL bio, we create an SSL object separately, setup its parameters, and then assign it to the BIO:
https://github.com/eclipse-openj9/openj9/blob/aba3173591037fb8aa8a5e9c4e4956a31074ae29/runtime/compiler/runtime/Listener.cpp#L162-L179

However, looking at the code of `BIO_new_ssl`, it already creates its own `SSL` object:

https://github.com/openssl/openssl/blob/c6fcd88fa030da8322cf27aff95376512f41faff/ssl/bio_ssl.c#L460-L478
```
BIO *BIO_new_ssl(SSL_CTX *ctx, int client)
{
    BIO *ret;
    SSL *ssl;


    if ((ret = BIO_new(BIO_f_ssl())) == NULL)
        return NULL;
    if ((ssl = SSL_new(ctx)) == NULL) {
        BIO_free(ret);
        return NULL;
    }
    if (client)
        SSL_set_connect_state(ssl);
    else
        SSL_set_accept_state(ssl);


    BIO_set_ssl(ret, ssl, BIO_CLOSE);
    return ret;
}
```

So when we replace the internal `SSL` with our own, it produces a memory leak.

The correct thing to do seems to be initializing `BIO` first, and then using `BIO_get_ssl` (or equivalent call to `BIO_ctrl`) to get the `SSL` object, and then setup its parameters. This should be done both on client and server sides.

When I implemented the above change, the crashes go away. However, it's not clear to me which objects need to be freed at stream termination, and in what order. 
`BIO_free_all` definitely needs to be called. But I haven't determined yet if that will also free its `SSL` object or if we need to call `SSL_free` separately.
